File: Day20-25/23.py
from collections import defaultdict


class VM:
    def __init__(self, instructions, registers=None):
        self.instructions = instructions
        self.pc = 0

        if registers is None:
            registers = defaultdict(int)
        self.registers = registers
        self.num_mul_ops = 0

    def _get_value(self, operand):
        try:
            return int(operand)
        except:
            return self.registers[operand]

    def run_instruction(self):
        inst = instructions[self.pc]
        if inst[0] == 'set':
            self.registers[inst[1]] = self._get_value(inst[2])
        elif inst[0] == 'sub':
            self.registers[inst[1]] -= self._get_value(inst[2])
        elif inst[0] == 'mul':
            self.num_mul_ops += 1
            self.registers[inst[1]] *= self._get_value(inst[2])
        elif inst[0] == 'jnz':
            if self._get_value(inst[1]) != 0:
                self.pc += self._get_value(inst[2])
                return
        self.pc += 1


def part2(instructions):
    start, end, skip = 0, 0, 0
    for inst in instructions:
        if inst[0] == 'set' and inst[1] == 'b':
            start = int(inst[2]) * 100 + 100000
        elif inst[0] == 'sub' and inst[1] == 'c':
            end = start - int(inst[2])
        elif inst[0] == 'sub' and inst[1] == 'b' and end:
            skip = -int(inst[2])

    h = 0
    for x in range(start, end+1, skip):
        for j in range(2, int(x**0.5)):
            if x % j == 0:
                h += 1
                break
    return h

if __name__ == '__main__':
    with open('23.txt') as f:
        instructions = [l.split() for l in f.readlines()]
    P1 = VM(instructions)
    while P1.pc < len(instructions):
        P1.run_instruction()
    print("Part 1: {}".format(P1.num_mul_ops))

    print("Part 2: {}".format(part2(instructions)))

File: Day20-25/23.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strconv"
	"strings"
)

type VM struct {
	instructions *[][]string
	pc           int64
	registers    map[string]int64

	mulOperations int
}

func (vm *VM) getValue(token string) int64 {
	val, err := strconv.Atoi(token)

	if err == nil {
		return int64(val)
	}

	if _, ok := vm.registers[token]; !ok {
		vm.registers[token] = 0
	}
	return vm.registers[token]
}

func (vm *VM) runInstruction() {
	inst := (*vm.instructions)[vm.pc]

	switch inst[0] {
	case "set":
		vm.registers[inst[1]] = vm.getValue(inst[2])
	case "sub":
		prev := vm.getValue(inst[1])
		vm.registers[inst[1]] = prev - vm.getValue(inst[2])
	case "mul":
		prev := vm.getValue(inst[1])
		vm.registers[inst[1]] = prev * vm.getValue(inst[2])
		vm.mulOperations++
	case "jnz":
		if vm.getValue(inst[1]) != 0 {
			vm.pc += vm.getValue(inst[2])
			return
		}
	default:
		panic("Invalid operation")
	}
	vm.pc++
}

func NewVM(instructions *[][]string, registers map[string]int64, part1 bool) VM {
	return VM{
		instructions: instructions,
		pc:           0,
		registers:    registers,

		mulOperations: 0,
	}
}

func part2(instructions *[][]string) int {
	var start, end, skip int
	for _, inst := range *instructions {
		if inst[0] == "set" && inst[1] == "b" {
			val, _ := strconv.Atoi(inst[2])
			start = val*100 + 100000
		} else if inst[0] == "sub" && inst[1] == "c" {
			val, _ := strconv.Atoi(inst[2])
			end = start - val
		} else if inst[0] == "sub" && inst[1] == "b" && end != 0 {
			val, _ := strconv.Atoi(inst[2])
			skip = -val
		}
	}

	h := 0
	for x := start; x <= end; x += skip {
		sqrtX := int(math.Sqrt(float64(x)))
		for j := 2; j <= sqrtX; j++ {
			if x%j == 0 {
				h++
			}
		}
	}
	return h
}

func main() {
	dat, _ := ioutil.ReadFile("./23.txt")
	lines := strings.Split(string(dat), "\n")

	instructions := make([][]string, len(lines))
	for i := 0; i < len(lines); i++ {
		instructions[i] = strings.Fields(lines[i])
	}

	P1 := NewVM(&instructions, make(map[string]int64), true)
	for P1.pc < int64(len(instructions)) {
		P1.runInstruction()
	}
	fmt.Printf("Part 1: %d\n", P1.mulOperations)

	fmt.Printf("Part 2: %d\n", part2(&instructions))
}
